python/test/unit/language/test_core.py

```diff
@@ -4329,30 +4329,105 @@ def mul_add(data):
 
     @triton.jit
-    def kernel(A, B, C):
-        tl.store(C, FUNC(tl.load(A), tl.load(B), propagate_nan=PROPAGATE_NAN))
-
-    kernel = patch_kernel(kernel, {'FUNC': func, 'PROPAGATE_NAN': propagate_nan})
+    def kernel(A, B, C, propagate_nan: tl.constexpr, func: tl.constexpr):
+        if func == 'clamp':
+            tl.store(
+                C,
+                getattr(tl, func)(tl.load(A), -tl.load(B), tl.load(B),
+                                  propagate_nan=getattr(tl.PropagateNan, propagate_nan)))
+        else:
+            tl.store(C,
+                     getattr(tl, func)(tl.load(A), tl.load(B), propagate_nan=getattr(tl.PropagateNan, propagate_nan)))
 
     for mode in ['A', 'B', 'both']:
-        A = torch.randn((1, ), device='cuda', dtype=torch.float32)
+        A = torch.randn((1, ), device='cuda', dtype=getattr(torch, dtype))
         if mode == 'A' or mode == 'both': A[0] = torch.nan
-        B = torch.randn((1, ), device='cuda', dtype=torch.float32)
+        B = torch.randn((1, ), device='cuda', dtype=getattr(torch, dtype))
         if mode == 'B' or mode == 'both': B[0] = torch.nan
-        C = torch.zeros_like(A, device='cuda', dtype=torch.float32)
-        kernel[(1, )](A, B, C)
+        C = torch.zeros_like(A, device='cuda', dtype=getattr(torch, dtype))
+        kernel[(1, )](A, B, C, propagate_nan, func)
 
-        if mode == 'both' or eval(propagate_nan) == tl.PropagateNan.ALL:
+        if mode == 'both' or propagate_nan == 'ALL':
             assert torch.isnan(C[0])
         else:
             assert not torch.isnan(C[0])
 
 
+# -----------------------
+# test clamp
+# -----------------------
+
+
+@pytest.mark.parametrize("dtype", ['float16', 'float32'])
+def test_clamp(dtype):
+
+    @triton.jit
+    def kernel(x_ptr, min_ptr, max_ptr, out_ptr, ref_ptr, N, BLOCK_SIZE: tl.constexpr):
+
+        off = tl.arange(0, BLOCK_SIZE)
+        mask = off < N
+        x = tl.load(x_ptr + off, mask=mask)
+        min = tl.load(min_ptr + off, mask=mask)
+        max = tl.load(max_ptr + off, mask=mask)
+        out = out_ptr + off
+        ref = ref_ptr + off
+
+        tl.store(out, tl.clamp(x, min, max), mask=mask)
+        ref_val = tl.minimum(tl.maximum(x, min), max)
+        tl.store(ref, ref_val, mask=mask)
+
+    size = 128
+
+    x = torch.randn((size, ), device='cuda', dtype=getattr(torch, dtype))
+    a = torch.randn((size, ), device='cuda', dtype=getattr(torch, dtype))
+    b = torch.randn((size, ), device='cuda', dtype=getattr(torch, dtype))
+    min = torch.min(a, b)
+    max = torch.max(a, b)
+    out = torch.zeros_like(x, device='cuda', dtype=getattr(torch, dtype))
+    ref = torch.zeros_like(x, device='cuda', dtype=getattr(torch, dtype))
+
+    kernel[(size, )](x, min, max, out, ref, x.numel(), BLOCK_SIZE=size)
+
+    torch.testing.assert_close(out, ref)
+
+
+# Test for symmetric clamp(x, -limit, limit), as it may go through optimized
+# codegen in the backends
```

Comment:
you could also check that the ptx has the optimized instruction, see example:
https://github.com/openai/triton/blob/3a95ce666ba2a62be71207a3c2f9babc754e85ff/python/test/unit/language/test_core.py#L2367

If not we would probably want a lit test to make sure the pattern works?

Reply:
This was intended to be a functional test to see if the results are correct in the pattermatched case (since I knew codegen is different). I agree lit test is warranted here, I'll add it.

Reply:
right that makes sense, what I'm saying is that you can do two in one and have the test that checks the result + check that the ptx has the instruction to make sure the optimization happened